### PR TITLE
#26707: fix matmul m padding

### LINKED ISSUE #26707: Input for matmul results in device side error
Run the following as a python program:
```
import torch    
import ttnn    
  
# Open device    
device = ttnn.open_device(device_id=0)    
  
# Create input tensors (keep the same random data for comparison)  
torch_a = torch.randn((1, 1, 64, 2880), dtype=torch.bfloat16)  
torch_b = torch.randn((1, 1, 2880, 2880), dtype=torch.bfloat16)  
  
tensor_a = ttnn.from_torch(    
    torch_a,    
    layout=ttnn.TILE_LAYOUT,    
    dtype=ttnn.bfloat8_b,  
    device=device,  
    memory_config=ttnn.L1_MEMORY_CONFIG  
)
tensor_b = ttnn.from_torch(    
    torch_b,    
    layout=ttnn.TILE_LAYOUT,    
    dtype=ttnn.bfloat8_b,  
    device=device,  
    memory_config=ttnn.L1_MEMORY_CONFIG  
)

compute_kernel_config_HiFi4 = ttnn.WormholeComputeKernelConfig(
    math_fidelity=ttnn.MathFidelity.HiFi4,
    math_approx_mode=False,
    fp32_dest_acc_en=True,
    packer_l1_acc=True,
)

program_config = ttnn.MatmulMultiCoreReuseMultiCast1DProgramConfig(    
    #compute_with_storage_grid_size=(13, 10), 
    compute_with_storage_grid_size=(8, 8), 
    in0_block_w=1,
    out_subblock_h=1,    
    out_subblock_w=1,    
    per_core_M=6,   
    per_core_N=22, 
    fuse_batch=True,    
    fused_activation=None,    
    mcast_in0=True,    
)

print(tensor_a.shape, tensor_b.shape)
# Perform matrix multiplication    
result = ttnn.matmul(
    tensor_a,
    tensor_b,
    program_config=program_config,
    compute_kernel_config=compute_kernel_config_HiFi4,
    memory_config=ttnn.L1_MEMORY_CONFIG,
    dtype=ttnn.bfloat16,
)

output = ttnn.to_torch(result)  
expected = torch.matmul(torch_a, torch_b)  
  
pcc = ttnn.pearson_correlation_coefficient(expected, output)
ttnn.close_device(device)
print(f"PCC {pcc}")
assert pcc >= 0.99  

```
Output:
```
2025-08-12 18:26:39.680 | critical |          Always | Device 0 worker core(x= 0,y= 0) virtual(x=18,y=18): ncrisc using noc1 tried to unicast read 1088 bytes to local L1[0x019260] from Tensix core w/ virtual coords (x=20,y=19) L1[addr=0x0016e440] (NOC target address overflow). (assert.hpp:107)
terminate called after throwing an instance of 'std::runtime_error'
  what():  TT_THROW @ /localdev/bbradel/tt-metal/tt_metal/impl/debug/watcher_device_reader.cpp:627: tt::exception
info:
Device 0 worker core(x= 0,y= 0) virtual(x=18,y=18): ncrisc using noc1 tried to unicast read 1088 bytes to local L1[0x019260] from Tensix core w/ virtual coords (x=20,y=19) L1[addr=0x0016e440] (NOC target address overflow).
backtrace:
 --- /localdev/bbradel/tt-metal/build_Debug/lib/_ttnncpp.so(+0x1c19e3b) [0x7fa8903bde3b]
 --- tt::tt_metal::WatcherDeviceReader::DumpNocSanitizeStatus(tt::umd::CoreDescriptor&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, mailboxes_t const*, int)
 --- tt::tt_metal::WatcherDeviceReader::DumpCore(tt::umd::CoreDescriptor&, bool)
```

On WH N150

### PR DESCRIPTION
**Pull request by bbradelTT**
### PR Category

Closes #26707 

M-direction padding when M < per_core_M was not set up properly.

This is being fixed and a test added.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:bbradel-26707_mm_m_padding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:bbradel-26707_mm_m_padding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:bbradel-26707_mm_m_padding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:bbradel-26707_mm_m_padding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:bbradel-26707_mm_m_padding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:bbradel-26707_mm_m_padding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:bbradel-26707_mm_m_padding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:bbradel-26707_mm_m_padding)

### CODE REVIEW

File: tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul.py [hunk 1]
```diff
@@ -453,3 +453,61 @@ def test_matmul_transpose_a_fuse_batch(device, batch, m, k, n, program_config):
     assert_numeric_metrics(
         torch_out, output, check_allclose=False, check_frobenius=False, pcc_threshold=0.999, check_ulp=False
     )
+
+
+def test_matmul_m_direction_padding(device):
+    # Create input tensors (keep the same random data for comparison)
+    torch_a = torch.randn((1, 1, 64, 2880), dtype=torch.bfloat16)
+    torch_b = torch.randn((1, 1, 2880, 2880), dtype=torch.bfloat16)
+
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul.py:463]: Added

File: tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul.py [hunk 2]
```diff
@@ -453,3 +453,61 @@ def test_matmul_transpose_a_fuse_batch(device, batch, m, k, n, program_config):
...
+        out_subblock_h=1,
+        out_subblock_w=1,
+        per_core_M=6,
+        per_core_N=22,
+        fuse_batch=True,
+        fused_activation=None,
+        mcast_in0=True,
+    )
+
+    print(tensor_a.shape, tensor_b.shape)
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul.py]: Removed

File: tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul.py [hunk 3]
```diff
@@ -453,3 +453,61 @@ def test_matmul_transpose_a_fuse_batch(device, batch, m, k, n, program_config):
...
+        program_config=program_config,
+        compute_kernel_config=compute_kernel_config_HiFi4,
+        memory_config=ttnn.L1_MEMORY_CONFIG,
+        dtype=ttnn.bfloat16,
+    )
+
+    output = ttnn.to_torch(result)
+    expected = torch.matmul(torch_a, torch_b)
+
+    pcc = ttnn.pearson_correlation_coefficient(expected, output)
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul.py]: Removed

File: ttnn/cpp/ttnn/operations/matmul/device/factory/matmul_multicore_reuse_mcast_1d_program_factory.cpp [hunk 4]
```diff
@@ -901,6 +902,17 @@ MatmulMultiCoreReuseMcast1DProgramFactory::shared_variables_t process_mcast_in0_
     uint32_t last_block_padded_block_tiles_w_skip =
         (out_subblock_w * out_subblock_h) * (out_block_w / out_subblock_w - last_block_num_nonzero_subblocks_w);
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/matmul/device/factory/matmul_multicore_reuse_mcast_1d_program_factory.cpp:921]: Added.

**Review by vsureshTT — APPROVED** [review]
